### PR TITLE
Merge ozan/kickoff-root-operations into withdrawal

### LINKED ISSUE #225: Fix operator related binaries
### Proposal Description

Since now we have multiple operators, some of the binaries need changes.

### PR DESCRIPTION
**Pull request by ozankaymak**
Closes #225 and #219.

### CODE REVIEW

File: core/src/verifier.rs [hunk 1]
```diff
@@ -146,91 +160,164 @@ where
...
                 &self.config.operators_xonly_pks[i],
             )?;
+
+            // Check if for each operator the address of the kickoff_utxo is correct TODO: Maybe handle the possible errors better
+            let musig2_and_operator_script =
+                script_builder::create_musig2_and_operator_multisig_script(
+                    &self.nofn_xonly_pk,
+                    &self.config.operators_xonly_pks[i],
+                );
+            let (musig2_and_operator_address, _) = TransactionBuilder::create_taproot_address(
```
Comment (ekrembal) [inline core/src/verifier.rs:176]: Lets have a create_kickoff_address function in transaction_builder

File: core/src/verifier.rs [hunk 2]
```diff
@@ -146,91 +160,164 @@ where
...
+            .collect::<Vec<_>>();
+        // let root: bitcoin::hashes::sha256::Hash = bitcoin::merkle_tree::calculate_root(
+        //     kickoff_utxos
+        //         .iter()
+        //         .map(|utxo| Hash::from_byte_array(sha256_hash!(utxo.to_vec().as_slice()))),
+        // )
+        // .unwrap();
+        // let root_bytes: [u8; 32] = *root.as_byte_array();
+
+        self.db
```
Comment (ekrembal) [inline core/src/verifier.rs:230]: We should save aggNonces first so that we can partialSign

File: core/src/verifier.rs [hunk 3]
```diff
@@ -146,91 +160,164 @@ where
...
     }
 
     /// verify burn txs are signed by verifiers
     /// sign operator_takes_txs
-    async fn burn_txs_signed_rpc(
+    async fn burn_txs_signed(
         &self,
         deposit_outpoint: OutPoint,
         _burn_sigs: Vec<schnorr::Signature>,
+        _slash_or_take_sigs: Vec<schnorr::Signature>,
```
Comment (ekrembal) [inline core/src/verifier.rs:317]: We should also verify these slash_or_take_sigs

**Review by ekrembal — CHANGES_REQUESTED** [review]

File: core/src/transaction_builder.rs [hunk 4]
```diff
@@ -318,38 +382,67 @@ impl TransactionBuilder {
...
+        let move_txins = TransactionBuilder::create_tx_ins(vec![deposit_outpoint]);
+        let move_txout = TxOut {
+            value: Amount::from_sat(BRIDGE_AMOUNT_SATS)
+                - Amount::from_sat(MOVE_TX_MIN_RELAY_FEE)
+                - anyone_can_spend_txout.value,
+            script_pubkey: musig2_address.script_pubkey(),
+        };
+
+        let move_tx_simple =
+            TransactionBuilder::create_btc_tx(move_txins, vec![move_txout, anyone_can_spend_txout]);
```
[the remark this answers is not in the document]
Comment (ozankaymak) [reply core/src/transaction_builder.rs:404]: Done

File: core/src/transaction_builder.rs [hunk 5]
```diff
@@ -318,38 +382,67 @@ impl TransactionBuilder {
...
-        let scripts = vec![vec![relative_timelock_script]];
-        let taproot_spend_infos = vec![slash_or_take_address.1];
+        let prevouts = vec![kickoff_utxo.txout.clone()];
+        let scripts = vec![vec![relative_timelock_script.clone()]];
         TxHandler {
             tx,
             prevouts,
             scripts,
-            taproot_spend_infos,
+            taproot_spend_infos: vec![],
```
[the remark this answers is not in the document]
Comment (ozankaymak) [reply core/src/transaction_builder.rs:445]: Done

File: core/src/verifier.rs (hunk 1 above)
Comment (ozankaymak) [reply core/src/verifier.rs:176]: Done

File: core/src/verifier.rs (hunk 3 above)
Comment (ozankaymak) [reply core/src/verifier.rs:317]: Done

File: core/src/verifier.rs [hunk 6]
```diff
@@ -119,20 +128,25 @@ where
...
+    /// do not forget to add tweak when signing since this address has n_of_n as internal_key
+    /// and operator_timelock as script.
     async fn operator_kickoffs_generated(
         &self,
         deposit_outpoint: OutPoint,
         kickoff_utxos: Vec<UTXO>,
         operators_kickoff_sigs: Vec<secp256k1::schnorr::Signature>, // These are not transaction signatures, rather, they are to verify the operator's identity.
-        agg_nonces: Vec<MuSigAggNonce>,
-    ) -> Result<Vec<MuSigPartialSignature>, BridgeError> {
+        agg_nonces: (Vec<MuSigAggNonce>, Vec<MuSigAggNonce>), // First vector is for the slash_or_take_txs, the second vector is for the burn_txs.
```
[the remark this answers is not in the document]
Comment (ozankaymak) [reply core/src/verifier.rs:142]: Done

**Review by ekrembal — APPROVED** [review]

**Review by ekrembal — APPROVED** [review]

File: core/src/transaction_builder.rs [hunk 7]
```diff
@@ -318,38 +382,67 @@ impl TransactionBuilder {
...
+        let relative_timelock_script =
+            script_builder::generate_relative_timelock_script(operator_xonly_pk, 200); // TODO: Change this 200 to a config constant
+        let (slash_or_take_address, _) = TransactionBuilder::create_taproot_address(
             &[relative_timelock_script.clone()],
             Some(*nofn_xonly_pk),
             network,
         );
-
+        let mut op_return_script: Vec<u8> = hex::decode(move_txid.to_string()).unwrap();
+        let op_return_idx: Vec<u8> = if operator_idx < 256 {
```
[the remark this answers is not in the document]
Comment (lemonpartee) [reply core/src/transaction_builder.rs:284]: This code snippet encodes variable `n` of type `usize` using least amount of bytes.
```rust
let usize_bytes = usize::BITS.div(8) as usize;
let bits = n.max(1).ilog2() + 1;
let len = bits.div_ceil(8) as usize;
let empty = usize_bytes - len;
let bytes = n.to_be_bytes();
let bytes = &bytes[empty..];
```

This code snippet would decode it.

```rust
let mut encoded = [0u8; 8];
encoded[empty..].copy_from_slice(bytes);
let decoded = usize::from_be_bytes(encoded);
```

File: core/src/errors.rs [hunk 8]
```diff
@@ -158,6 +158,12 @@ pub enum BridgeError {
     KickoffOutpointsNotFound,
     #[error("DepositInfoNotFound")]
     DepositInfoNotFound,
+
+    #[error("FromHexError: {0}")]
+    FromHexError(hex::FromHexError),
```
Comment (lemonpartee) [inline core/src/errors.rs:163]: You don't have to manually implement `From` if you use this syntax.
```suggestion
    FromHexError(#[from] hex::FromHexError),
```

File: core/src/servers.rs [hunk 9]
```diff
@@ -23,21 +24,21 @@ use traits::rpc::OperatorRpcServer;
...
         .await
     {
         Ok(s) => s,
         Err(e) => return Err(BridgeError::ServerError(e)),
     };
-
     let verifier = Verifier::new(rpc, config).await?;
 
-    let addr = match server.local_addr() {
+    let addr: std::net::SocketAddr = match server.local_addr() {
```
Comment (lemonpartee) [inline core/src/servers.rs:41]:
```suggestion
    let addr: std::net::SocketAddr = server.local_addr().map_err(|e| BridgeError::ServerError(e)|)?;
```

**Review by lemonpartee — COMMENTED** [review]
Left a couple nits

**Pull request merged**